### DISCUSSION TITLE
How can OLE2FRAME data be extracted?

### DISCUSSION THREAD

**Discussion by Fortunox** [mozman/ezdxf#1097] [Q&A]
Dear mozman,
first let me say thanks for ezdxf!
I'm searching for a posibility to extract the data of Excel tables which are stored in OLE2FRAMES within the dxf-files. I found out that OLE2FRAMES are managed with the DXFTagStorage class. But unfortunatelly I didn't find the way to extract the OLE2 data. I also have tried to find a hint in https://ezdxf.readthedocs.io.
Would it be possible to give me a short example how to extract the OLE2 data.
Thanks and best regards,
Fortunox

**Comment by Fortunox** [comment]
Many thanks mozman for this fast answer and your help!
I will have a closer look to that in the next few days and will give a feedback here!

**Comment by pincusz** [comment]
@mozman

Has anyone got this to work? I have a few OLE2Frame entities that should be excel tables but im really struggling to extract and get the data. I need to be able to turn it into a dataframe to compare to other excel data.

My process so far has been, extract the OLE2Frame entities. get the binary data using tag.code == 310. Then I get stuck. I have a list of the binary data, at least i think i do? When i try to convert it, I keep getting errors "Error: Non-hexadecimal digit found". To me it looks like the data from tags == 310 is byte strings. 

I am confused as to what the next step is, how can i go from this list of byte string data into getting what the excel table data looks like? ive never worked with OLEs or raw binary data like this before. Thanks for any advice!
![Capture](https://github.com/mozman/ezdxf/assets/171051038/98ab10df-fd10-4b6f-b61f-16a9c8ae1eab)

**Comment by mozman** [comment]
Updated code to work with newer versions of `ezdxf`.

OLE2FRAMES are stored as `OLE2Frame` instances for preservation, but if you have to extract data, it is possible:

```Python
import ezdxf


def main():
    doc = ezdxf.readfile("your.dxf")
    msp = doc.modelspace()
    for ole in msp.query('OLE2FRAME'):
        # get subclass as Tags() object
        tags = ole.acdb_ole2frame
        
        # for tag codes look: 
        # http://help.autodesk.com/view/OARX/2018/ENU/?guid=GUID-77747CE6-82C6-4452-97ED-4CEEB38BE960
        # DXF files have often missing data, therefor default to None, else DXFValueError() will raise
        # for missing data
        kind = tags.get_first_value(3, default=None)
        corner1 = tags.get_first_value(10, default=None)
        corner2 = tags.get_first_value(11, default=None)
        ole_type = tags.get_first_value(71, default=None)
        tile_mode = tags.get_first_value(72, default=None)
        # you get the idea
        
        # extract binary encoded data as bytes: b'\xf0\xf0...' 
        bin_data = b"".join(tag.value for tag in tags if tag.code == 310)


if __name__ == "__main__":
    main()
```

**Reply by pincusz** [reply]
@mozman Thanks for the updated code, I have a follow up question.

- You mentioned that the OLE entities data is "preserved". Does this mean with the bin_data that is extracted from this code, I can use another library to "reverse engineer" or whatever to get it back into excel table format? My main goal is that I have excel tables as OLE entities in the DXF, I need to read the contents of these excel tables. So now that I have the OLE binary data, does that mean I can get the excel data?

**Reply by mozman** [reply]
You will get all the binary data stored in the DXF file. I don't know if that is enough to recover the Excel files.

Here is the DXF reference for the `OLE2FRAME` entity: https://help.autodesk.com/view/ACD/2022/ENU/?guid=GUID-77747CE6-82C6-4452-97ED-4CEEB38BE960
